modules/tobira/src/main/java/org/opencastproject/tobira/impl/TobiraApi.java

```diff
@@ -103,79 +138,154 @@ public void setSearchService(SearchService service) {
+              return true;
             }
+
+            final Date lastSeriesModifiedDate = rawSeries.get(rawSeries.size() - 1)
+                .getModifiedDate();
+            return !event.getModified().after(lastSeriesModifiedDate);
           })
+          .map(event -> new Item(event));
+
+      final Stream<Item> seriesItems = rawSeries.stream()
+          .limit(preferredAmount)
+          .map(series -> new Item(series));
+
+
+      // Combine series and events into one combined list and sort it.
+      //
+      // The sorting is, again, not for correctness, because consumers of this API need to be
+      // able to deal with that. However, sorting this here will result in fewer temporary objects
+      // or invalid states in the consumer.
+      final ArrayList<Item> items = Stream.concat(eventItems, seriesItems)
           .collect(Collectors.toCollection(ArrayList::new));
+      items.sort(Comparator.comparing(item -> item.getModifiedDate()));
+
+
+      // Obtain information to allow Tobira to plan the next harvesting request.
+      boolean hasMore = hasMoreEvents || hasMoreSeriesInRange;
+      final Date includesItemsUntilRaw;
+      if (!hasMoreEvents && !hasMoreSeriesInRange) {
+        // All events and series up to now have been harvested.
+        includesItemsUntilRaw = new Date();
+      } else if (!hasMoreEvents && hasMoreSeriesInRange) {
+        // `rawEvents` contains all events that currently exist. Our response won't contain the ones
+        // that have a modified date after the one of the last raw series. But that means that we
+        // will return all series and events until the last raw series.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      } else if (hasMoreEvents && !hasMoreSeriesInRange) {
+        // There are more events, but no additional series in the range from `since` to the modified
+        // date of the last raw event. So we know there are no other events or series before the
+        // last raw events.
+        includesItemsUntilRaw = rawEvents[rawEvents.length - 1].getModified();
+      } else {
+        // There are more events and more series in the given range. In theory, this would be
+        // `Math.min()` of the last raw event and last raw series. However, since `hasMoreEvents`
+        // is set, we only loaded series with modified dates smaller than the last raw event's
+        // modified date. That means, that the modified date of the last raw series is always
+        // smaller than that of the last raw event.
+        includesItemsUntilRaw = rawSeries.get(rawSeries.size() - 1).getModifiedDate();
+      }
+
+      // The `includesItemsUntil` we return has to be at least `TIME_BUFFER_SIZE` in the past. See
+      // the constant's documentation for more information on that.
+      final long includesItemsUntil = Math.min(
+          includesItemsUntilRaw.getTime(),
+          new Date().getTime() - TIME_BUFFER_SIZE
+      );
```

Comment:
It's all documentation, but instead of modifying the value up until which this request includes data, it might be less confusing a) to modify the request parameter or b) expect the client to do that. That also makes this code easier since it would just become:
```suggestion
      final long includesItemsUntil = includesItemsUntilRaw.getTime();
```

Reply:
I'm not sure I understand. The client just takes the `includesItemsUntil` value and uses it as the `since` value for the next harvest request. And I would say the `TIME_BUFFER_SIZE` should be decided by the Opencast side. So I'm not entirely sure what you're suggesting here.